File: 算法/查找/二分查找/剑指offer11旋转数组的最小数字.py
from typing import List


class Solution:
    def minArray(self, numbers: List[int]) -> int:
        if not numbers:
            return None

        if len(numbers) == 1:
            return numbers[0]

        left = 0
        right = len(numbers) - 1
        while left < right:
            mid = left + (right - left) // 2
            if numbers[mid] < numbers[right]:
                right = mid
            elif numbers[mid] > numbers[right]:
                left = mid + 1  # 这是不会越界的
            else:
                # 无法判断在哪里
                right = right - 1

        return numbers[left]

File: 算法/查找/二分查找/剑指offer11旋转数组的最小数字.cpp
#include <vector>
#include <algorithm>
#include <iostream>

using namespace std;

class Solution
{
public:
    int minArray(vector<int> &numbers)
    {
        if (!(&numbers) || numbers.size() == 0)
        {
            return 0;
        }
        if (numbers.size() == 1)
        {
            return numbers[0];
        }

        vector<int>::size_type left = 0;
        vector<int>::size_type right = numbers.size() - 1;

        while (left < right)
        {
            auto mid = left + (right - left) / 2;
            if (numbers[mid] < numbers[right])
            {
                right = mid;
            }
            else if (numbers[mid] > numbers[right])
            {
                left = mid + 1;
            }
            else
            {
                right = right - 1;
            }
        }

        return numbers[left];
    }
};